compiler/noirc_frontend/src/elaborator/statements.rs

```diff
@@ -279,13 +281,21 @@ impl<'context> Elaborator<'context> {
             self.push_err(ResolverError::LoopInConstrainedFn { span });
         }
 
-        self.nested_loops += 1;
+        let old_loop = std::mem::take(&mut self.current_loop);
+        self.current_loop = Some(Loop { is_for: false, has_break: false });
         self.push_scope();
 
         let (block, _block_type) = self.elaborate_expression(block);
 
         self.pop_scope();
-        self.nested_loops -= 1;
+
+        if !in_constrained_function {
+            let last_loop =
+                std::mem::replace(&mut self.current_loop, old_loop).expect("Expected a loop");
+            if !last_loop.has_break {
+                self.push_err(ResolverError::LoopWithoutBreak { span });
+            }
+        }
```

Comment:
Why guard this with a `if !in_constrained_function`?
Even if the user writes a `loop {}` in a function they accidentally let be constrained I don't see the harm in issuing a loop without break error. I also think because the `std::mem::replace` is still here it opens up more chances for it to get out of sync if we're in a constrained function.

Reply:
My thinking was that the user already got an error for the `loop`, so giving another error might be odd (if you can't have a loop here, does it matter whether it has a break or not?). But I'm fine either way, so let me know.

Reply:
I'd prefer the second error because there are still two things wrong with the code. I tend to disagree with omitting any errors in general since it makes fixing code more difficult (need to fix the first issue, then save and run the compiler again to see the second). In this case I think it could be the case the user wanted a `loop` but forgot to make their function unconstrained first so they'll need to fix both errors anyway.

Reply:
Done! Also the thing you said about `std::mem::replace` is true, I meant for `last_loop` to be out of that `if` but I ended up putting it inside it.